packages/components/src/tools-panel/test/index.js

```diff
@@ -348,6 +349,237 @@ describe( 'ToolsPanel', () => {
+			// the optional menu item group.
+			optionalItem = screen.queryByRole( 'menuitemcheckbox', {
+				name: 'Show Linked',
+			} );
+			expect( optionalItem ).not.toBeInTheDocument();
+		} );
+
+		it( 'should handle conditionally rendered default control', async () => {
+			const conditionalControlProps = {
+				attributes: { value: false },
+				hasValue: jest.fn().mockImplementation( () => {
+					return !! conditionalControlProps.attributes.value;
+				} ),
+				label: 'Conditional',
+				onDeselect: jest.fn(),
+				onSelect: jest.fn(),
+			};
+
+			const TestPanel = () => (
+				<ToolsPanel { ...defaultProps }>
+					<ToolsPanelItem
+						{ ...altControlProps }
+						isShownByDefault={ true }
+					>
+						<div>Default control</div>
+					</ToolsPanelItem>
+					{ !! altControlProps.attributes.value && (
+						<ToolsPanelItem
+							{ ...conditionalControlProps }
+							isShownByDefault={ true }
+						>
+							<div>Conditional control</div>
+						</ToolsPanelItem>
+					) }
+				</ToolsPanel>
+			);
+
+			const { rerender } = render( <TestPanel /> );
+
+			// The conditional control should not yet be rendered.
+			let conditionalItem = screen.queryByText( 'Conditional control' );
+			expect( conditionalItem ).not.toBeInTheDocument();
+
+			// The conditional control should not yet appear in the default controls
+			// menu group.
+			openDropdownMenu();
+			let menuGroups = screen.getAllByRole( 'group' );
+			let defaultItem = within( menuGroups[ 0 ] ).queryByText(
+				'Conditional'
+			);
+			expect( defaultItem ).not.toBeInTheDocument();
+
+			// Simulate the main control having a value set which will now
+			// render the new default control into the ToolsPanel.
+			altControlProps.attributes.value = true;
+
+			rerender( <TestPanel /> );
+
+			// The conditional control should now be rendered and included in
+			// the panel's menu.
+			conditionalItem = screen.getByText( 'Conditional control' );
+			expect( conditionalItem ).toBeInTheDocument();
+
+			// The conditional control should now appear in the default controls
+			// menu group.
+			menuGroups = screen.getAllByRole( 'group' );
+
+			// The new default control item for the Conditional control should
+			// be within the first menu group.
+			defaultItem = within( menuGroups[ 0 ] ).getByText( 'Conditional' );
+			expect( defaultItem ).toBeInTheDocument();
+		} );
+	} );
+
+	describe( 'registration of panel items', () => {
+		it( 'should register and deregister items when switching between panels', () => {
+			// This test adds a `panelId` prop to two panels, to simulate
+			// the registration and de-registration of panelItems when
+			// switching between panels and test for any issues.
+			//
+			// When you switch from panel A to panel B, only panel A items
+			// should be deregistered, and only panel B items should be registered.
+			//
+			// See: https://github.com/WordPress/gutenberg/pull/36588
```

Comment:
I wonder if the direct references here to two panels obscure some of the real-world cases we are trying to guard against here.

Take a block support panel that renders the tools panel items via SlotFill. There is only a single `ToolsPanel` in play. When the block selection changes, so does that component's `panelId` prop and it rerenders. The fills containing injected controls update separately which is why we still have items belonging under the old `panelId` and needing to deregister them.

Maybe we could tweak this a bit to communicate that? 

Feel free to ignore the suggestion but it could help us in future to quickly grok exactly why we added the test in the first place.

```suggestion
		it( 'should register and deregister items when panelId changes', () => {
			// This test simulates switching block selections which cause the
			// `ToolsPanel` to rerender with a new `panelId` and as such, have
			// to register or deregister appropriate `ToolsPanelItem` children.
			//
			// When the `panelId` changes, only items matching the new ID should
			// register themselves while those for the old `panelId` deregister.
			//
			// See: https://github.com/WordPress/gutenberg/pull/36588
```

Reply:
I think this is exactly what I was _trying_ to describe, but using "a different panel" to describe "a different panelId" is imprecise and made it unclear. I like the language you've suggested much better, I think it communicates the exact problem. Done, thank you for the suggestion!